### PR TITLE
Readme renderer update

### PR DESCRIPTION
**Pull request by GuusVink**
Included general information of the repo
+
formed the readme template outside of the JS file (for ease of markdown editing)
+
included all types that are visited by the analyzer and thus included in the generated JSON (along with hyperrefs to the JavaParser documentation) -> this addresses the 'issue' of the current state of the analyzer (i.e., what language constructs are supported)

### CODE REVIEW

File: ldj-readme-renderer/README_template.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+capable to handle multilingual projects. One aspect of this is to make LivingDocumentation
+available for the Java environment; which is the purpose of this repository.
+
+Within the aspect of making LivingDocumentation available for the Java environment, two features
+can be identified (which are copied from [LivingDocumentation][ldoc] description):
+
+* [**Analyzer**](#analyzer): A tool to analyze Java projects.
+* [**Libraries**](#support-libraries): Assists in generating applications that can create plain text files such as MarkDown, AsciiDoc, PlantUML, Mermaid, and more.
+
+> **Note**: As this repository is currently association to a University assignment, it is unfortunately not possible to directly contribute to this project.
```
Comment (wandernauta) [inline ldj-readme-renderer/README_template.md:25]: Aww, are you sure about this? I would think we'd gladly take patches, if people have them.

File: ldj-readme-renderer/render.js [hunk 2]
```diff
@@ -8,45 +8,57 @@
...
-
-## Analyzer
-
-The analyzer parses Java projects into Living Documentation JSON files that can
-be rendered by Living Documentation renderers. It accepts the following command
-line options:
+const json = JSON.parse(fs.readFileSync(process.argv[2]));
 
+// Section for rendering the command line options
+const CMD_LINE_TABLE_TEMPLATE = `
```
Comment (wandernauta) [inline ldj-readme-renderer/render.js:16]: Why is this here, and not in `README_template.md`?

File: ldj-readme-renderer/render.js [hunk 3]
```diff
@@ -8,45 +8,57 @@
...
   }
 }
 
-const json = JSON.parse(fs.readFileSync(process.argv[2]));
-fs.writeFileSync(OUTPUT, TEMPLATE.replace('~', () => render(json)));
+const cmdOptionsTable = CMD_LINE_TABLE_TEMPLATE.replace('~', () => renderCmdOptions(json));
+
+// Section rendering the implemented visited methods
+
+const JAVAPARSER_DOC_SITE = 'https://javadoc.io/static/com.github.javaparser/javaparser-core/3.25.6/';
```
Comment (wandernauta) [inline ldj-readme-renderer/render.js:40]: Probably not a big difference, but we use version 3.25.1, not 3.25.6 (see analyzer-java's pom.xml).

File: ldj-readme-renderer/render.js [hunk 4]
```diff
@@ -8,45 +8,57 @@
...
+function getMarkdownLink(type) {
+  let splitted = type.split('.');
```
Comment (wandernauta) [inline ldj-readme-renderer/render.js:42]: Here and below, all except `finalReadme` can be `const` instead of `let`.

File: ldj-readme-renderer/render.js [hunk 5]
```diff
@@ -8,45 +8,57 @@
...
+}
+
+let visitor = json.filter((type) => type.FullName === 'com.infosupport.ldoc.analyzerj.AnalysisVisitor')[0];
+let parserTypes = visitor.Methods.filter((method) => method.Name === 'visit').map((method) => method.Parameters[0].Type);
+let implementedTypesList = parserTypes.map(getMarkdownLink).join(', ');
+
+// Combining everything
+
+try {
+  let readmeTemplate = fs.readFileSync('README_template.md', 'utf8');
```
Comment (wandernauta) [inline ldj-readme-renderer/render.js:55]: Suggest extracting `'README_template.md'` this as a constant and putting it near line 11.

**Review by wandernauta — COMMENTED** [review]
First round done, some general questions and some suggestions for the actual code.

File: ldj-readme-renderer/README_template.md [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+* [**Analyzer**](#analyzer): A tool to analyze Java projects.
+* [**Libraries**](#support-libraries): Assists in generating applications that can create plain text files such as MarkDown, AsciiDoc, PlantUML, Mermaid, and more.
+
+> **Note**: As this repository is currently association to a University assignment, it is unfortunately not possible to directly contribute to this project. 
+> Furthermore, some of the described functionalities may not yet be finalised/are subject to change.
+
+## Getting started
+
+The process of rendering documentation can be summarized as follows:
+1. **Analysis**: Run the Analyzer over the source code of your project. This creates a LivingDocumentation JSON file.
```
[the remark this answers is not in the document]
Comment (GuusVink) [reply ldj-readme-renderer/README_template.md:31]: The only disadvantage with that is that if the options change (either the literal spelling or usage) this would have to be changed manually as opposed to the table below.
Would be kind of undermining if as a consequence of this the readme would contain a contradiction somewhere in the future (the word undermining is maybe a bit extreme but you get the sentiment).

This is why I tried to keep the 'fixed parts' rather general. Curious how you see this

File: ldj-readme-renderer/README_template.md [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+It can be used either as a Java command line application or as a Maven plugin.
+
+### Command line interface
+The Java analyzer can be invoked directly as a plain Java application [analyzer-java](/analyzer-java). It accepts the following command
+line options:
+
+[~cmd-line-options]
+
+### Maven plugin
```
[the remark this answers is not in the document]
Comment (GuusVink) [reply ldj-readme-renderer/README_template.md:49]: Would a note containing something like 'at the moment the plugin is not published, but it can be installed manually' be a good solution?

File: ldj-readme-renderer/README_template.md [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+The analyzer can also be used as a Maven plugin (which is defined in [ldj-maven-plugin](/ldj-maven-plugin)).
+For an example on how to use the Maven plugin, see the `pom.xml` configuration in [ldj-maven-plugin-example](/ldj-maven-plugin-example).
+
+### State of analyzer
```
[the remark this answers is not in the document]
Comment (GuusVink) [reply ldj-readme-renderer/README_template.md:53]: Main reason was to address questions like _"What types of classes and relations can this tool model?"_. This gives a dynamic list of the java features that are covered by the analyzer. Relation to the JavaParser documentation is to give a description of the exact java feature (which is not generally deductible from our code I think). In the future this sort of documentation could be replaced by or linked to the JSON schema documentation; answering the above question.

File: ldj-readme-renderer/README_template.md (hunk 1 above)
Comment (GuusVink) [reply ldj-readme-renderer/README_template.md:25]: I am not sure about this as we are graded on how we work as a project group and the work we deliver should be our own. 
On one hand, it could be considered positive if we manage an open-source project correctly resulting in a better product for Info Support. On the other hand, other people should probably not take over our work.
But again, this is a guess from my side on how Petra would view this (maybe discuss this sometime with Petra and others?)

File: ldj-readme-renderer/render.js (hunk 2 above)
Comment (GuusVink) [reply ldj-readme-renderer/render.js:16]: Because the Javascript code filling the table is dependent on the exact form of this smaller template (i.e. #columns & headers)

File: ldj-readme-renderer/render.js (hunk 4 above)
Comment (GuusVink) [reply ldj-readme-renderer/render.js:42]: I am not that familiar with the JS code styling guidelines, so excuse me for my ignorance (:  But why would `const` be preferred if it does not semantically represent a constant?
And if using `const`, would you also not need to change the naming to capital snake case?

File: ldj-readme-renderer/README_template.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (wandernauta) [reply ldj-readme-renderer/README_template.md:49]: Sure! We'd have to include instructions on how to do that manual installation, and make sure that those indeed work.

File: ldj-readme-renderer/README_template.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (wandernauta) [reply ldj-readme-renderer/README_template.md:53]: Hmm, right - I think I get what you mean, but if this is the intention, the list is not accurate, right? It includes `CatchClause,` but we do not include information about catch clauses in the JSON. I think we actually can not do so, without breaking compatibility; so listing it as something we already do seems odd to me.

If the hyperlinks are currently not useful, but will be useful in the future, I would suggest removing them for now and adding them at that future point.

File: ldj-readme-renderer/README_template.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (wandernauta) [reply ldj-readme-renderer/README_template.md:31]: I get what you mean. Indeed, let's keep it like this.

File: ldj-readme-renderer/render.js (hunk 4 above)
Comment (wandernauta) [reply ldj-readme-renderer/render.js:42]: We didn't officially pick any JavaScript style guide, but if you want to be consistent with the existing code, I went with:

* mark all variables `const` where possible, to indicate we don't expect them to be reassigned
* give only "real" constants (in this case, hard-coded settings) shouty-case names

That happens to align with Google JavaScript style, see [here](https://google.github.io/styleguide/jsguide.html#features-use-const-and-let) and [here](https://google.github.io/styleguide/jsguide.html#naming-constant-names).

File: ldj-readme-renderer/README_template.md [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+As of right now, this project is linked to a student assignment originating from the University of
+Twente and InfoSupport. The aim of the project is to make the [LivingDocumentation][ldoc] project
+capable to handle multilingual projects. One aspect of this is to make LivingDocumentation
+available for the Java environment; which is the purpose of this repository.
+
+Within the aspect of making LivingDocumentation available for the Java environment, two features
+can be identified (which are copied from [LivingDocumentation][ldoc] description):
+
+* [**Analyzer**](#analyzer): A tool to analyze Java projects.
+* [**Libraries**](#support-libraries): Assists in generating applications that can create plain text files such as MarkDown, AsciiDoc, PlantUML, Mermaid, and more.
```
[the remark this answers is not in the document]
Comment (GuusVink) [reply ldj-readme-renderer/README_template.md:23]: I have now added a note that they are not included in the project yet. As they are relevant for understanding the entire project (LD in general) I have still included them in the general explanations

File: ldj-readme-renderer/README_template.md (hunk 1 above)
Comment (GuusVink) [reply ldj-readme-renderer/README_template.md:25]: I created a general section about this as discussed with Petra

File: ldj-readme-renderer/render.js (hunk 4 above)
Comment (GuusVink) [reply ldj-readme-renderer/render.js]: I have changed this, for the report, it may be also nice to mention this style guide in the QA

File: ldj-readme-renderer/README_template.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (GuusVink) [reply ldj-readme-renderer/README_template.md]: As talked about, right now this list is excluded. If in a later stadium it might be useful to again include this information (albeit in some other documentation artefact), we can revert (or copy from the inverse of ) commit `fff85fa0339fcb6100e516939d0fc3bac867e460` 

Furthermore, this readme now contains a sort of placeholder reference to the JSON schema documentation which is yet to be approved/merged

**Comment by GuusVink** [comment]
@wandernauta I have addressed all issues. Once you approve (and merge) this PR, please either add a reference to the JSON schema documentation in readme section 'State of analyzer' (if at that moment the JSON schema doc is merged with main). Or add a note to the JSON schema doc PR to insert a reference to that doc within the README template of this PR

File: ldj-readme-renderer/README_template.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (wandernauta) [reply ldj-readme-renderer/README_template.md:48]: That really works - just `mvn install` inside the plugin directory and off you go? That's neat.

I would have expected there to be more steps, to build and install the analyzer perhaps, but if you've tried it and confirmed that's not needed, even for people who haven't used this before, great.

File: ldj-readme-renderer/render.js [hunk 10]
```diff
@@ -8,45 +8,43 @@
...
   } else if (d instanceof Object) {
-    return Object.values(d).map(render).join('');
+    return Object.values(d).map(renderCmdOptions).join('');
   }
 }
 
-const json = JSON.parse(fs.readFileSync(process.argv[2]));
-fs.writeFileSync(OUTPUT, TEMPLATE.replace('~', () => render(json)));
+const cmdOptionsTable = CMD_LINE_TABLE_TEMPLATE.replace('~', () => renderCmdOptions(json));
+const visitor = json.filter((type) => type.FullName === 'com.infosupport.ldoc.analyzerj.AnalysisVisitor')[0];
```
Comment (wandernauta) [inline ldj-readme-renderer/render.js:38]: This now seems unused.

File: ldj-readme-renderer/render.js [hunk 11]
```diff
@@ -8,45 +8,43 @@
...
+
+// Combining everything
+try {
+  const readmeTemplate = fs.readFileSync(README_TEMPLATE, 'utf8');
+  const finalReadme = readmeTemplate.replace('[~cmd-line-options]', cmdOptionsTable);
+  fs.writeFileSync(OUTPUT, finalReadme);
+} catch (err) {
+  console.error(err);
```
Comment (wandernauta) [inline ldj-readme-renderer/render.js:46]: I think the default behavior for an uncaught exception is to write the error to the console - does catching and logging it explicitly here make an important difference?

If not, I'd suggest removing the try/catch. That also makes it consistent with the `readFileSync` of the JSON file, above.

**Review by wandernauta — COMMENTED** [review]
Did a second round, found some smaller things.

I'll keep an eye on #90 as well so we can link that as soon as it lands.

File: ldj-readme-renderer/README_template.md [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+This project is an extension of the [Living Documentation][ldoc] set of tools to
+Java. As part of the Living Documentation toolchain, it can be used to generate
+documentation and diagrams from source code so that documentation is always
+up-to-date.
+
+## State of the project
+
+As of right now, this project is linked to a student assignment originating from the University of
+Twente and InfoSupport. The aim of the project is to make the [LivingDocumentation][ldoc] project
```
[the remark this answers is not in the document]
Comment (GuusVink) [reply ldj-readme-renderer/README_template.md:15]: Weird. Fixed both

File: ldj-readme-renderer/render.js (hunk 10 above)
Comment (GuusVink) [reply ldj-readme-renderer/render.js:38]: Correct, missed that one

File: ldj-readme-renderer/README_template.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (GuusVink) [reply ldj-readme-renderer/README_template.md]: Note for future reference:  commit `ee5deea9df4c85eab988b5098042fb2d5573cc77` now also contains a deletion of a required statement for this feature

File: ldj-readme-renderer/render.js (hunk 11 above)
Comment (GuusVink) [reply ldj-readme-renderer/render.js:46]: For the purpose of this renderer does indeed not make a difference. Agreed, so adjusted

**Review by wandernauta — APPROVED** [review]
Looks good to me!
